panoptes_aggregation/batch_aggregation.py

```diff
@@ -0,0 +1,156 @@
+
+        self.cls_csv = cls_file
+        self.wf_csv = wf_file
+        return {'classifications': cls_file, 'workflows': wf_file}
+
+    def process_wf_export(self, wf_csv):
+        self.wf_df = pd.read_csv(wf_csv)
+        self.wf_maj_version = self.wf_df.query(f'workflow_id == {self.workflow_id}')['version'].max()
+        self.wf_min_version = self.wf_df.query(f'workflow_id == {self.workflow_id} & version == {self.wf_maj_version}')['minor_version'].max()
+        self.workflow_version = f'{self.wf_maj_version}.{self.wf_min_version}'
+        self.workflow_row = self.wf_df.query(f'workflow_id == {self.workflow_id} & minor_version == {self.wf_min_version}')
+        self.tasks = json.loads(self.workflow_row.iloc[0]['tasks'])
+        return self.wf_df
+
+    def process_cls_export(self, cls_csv):
+        cls_df = pd.read_csv(cls_csv)
+        self.cls_df = cls_df.query(f'workflow_version == {self.workflow_version}')
+        return self.cls_df
+
+    def connect_blob_storage(self):
+        connect_str = os.getenv('AZURE_STORAGE_CONNECTION_STRING')
+        self.blob_service_client = BlobServiceClient.from_connection_string(connect_str)
+        self.blob_service_client.create_container(name=self.id)
+
+    def upload_file_to_storage(self, container_name, filepath):
+        blob = filepath.split('/')[-1]
+        blob_client = self.blob_service_client.get_blob_client(container=container_name, blob=blob)
+        with open(file=filepath, mode="rb") as data:
+            blob_client.upload_blob(data, overwrite=True)
+
+    def upload_files(self):
+        self.connect_blob_storage()
+        reductions_file = f'{self.output_path}/{self.workflow_id}_reductions.csv'
+        self.upload_file_to_storage(self.id, reductions_file)
+        zipfile = make_archive(f'tmp/{self.id}', 'zip', self.output_path)
+        self.upload_file_to_storage(self.id, zipfile)
+
+    def update_panoptes(self):
+        # An Aggregation class can be added to the python client to avoid doing this manually
+        params = {'workflow_id': self.workflow_id, 'user_id': self.user_id}
+        response = Panoptes.client().get('/aggregations/', params=params)
+        fresh_etag = response[1]
+
+        Panoptes.client().put(
+            '/aggregations/',
+            etag=fresh_etag,
+            json={
+                'aggregations': {
+                    'uuid': self.id,
+                    'status': 'completed'
+                }
+            }
+        )
+
+    def _generate_uuid(self):
+        self.id = uuid.uuid4().hex
+
+    def _download_export(self, url, filepath):
+        http = urllib3.PoolManager()
+        r = http.request('GET', url, preload_content=False)
+        with open(filepath, 'wb') as out:
+            while True:
+                data = r.read(65536)
+                if not data:
+                    break
+                out.write(data)
+        r.release_conn()
+
+    def _connect_api_client(self):
```

Comment:
Since the client is just sending HTTP requests on demand, I figured there was nothing to close past letting python clean up after itself. "Connect" really just means "go get a bearer token" as far as the client is concerned.